columnq/src/table/arrow_ipc_stream.rs

```diff
@@ -0,0 +1,95 @@
+use std::sync::Arc;
+
+use datafusion::arrow;
+use datafusion::arrow::datatypes::Schema;
+use datafusion::arrow::record_batch::RecordBatch;
+use log::debug;
+
+use crate::error::ColumnQError;
+use crate::table::TableSource;
+
+pub async fn to_mem_table(
+    t: &TableSource,
+) -> Result<datafusion::datasource::MemTable, ColumnQError> {
+    debug!("inferring arrow table schema...");
+    let schema_ref: arrow::datatypes::SchemaRef = match &t.schema {
+        Some(s) => Arc::new(s.into()),
+        None => {
+            let schemas = partitions_from_table_source!(t, |mut r| {
```

Comment:
same here, we can get the schema from the partitions below to avoid extra round of IO.